elastic/beats
# Add per-process network data to system/process

## Proposed commit message

Addresses part of https://github.com/elastic/beats/issues/7461 by adding a series of fields that report network usage in bytes for a process.

This ended up being a tad more complex than I had first thought, as there's two moving pieces.

- A component that needs to operate in its own thread, listening to incoming packets and recording byte counts
- A watcher component needs to operate independently monitoring `/proc/net/tcp` and `proc/net/udp` to monitor for network connections, then match the connection tuple to an inode, which can then be mapped to a process.

The latter part is accomplished by using packetbeat's `ProcessWatcher` component. 

Getting this to work as-is has required a few tradeoffs:

- Using libpcap would introduce an other runtime dependency into metricbeat, so this uses `AF_PACKET`, which is linux-only. I'm not sure if there's a "clever" way to swap between the two, as the libpcap dependency is introduced when the package is imported.
- `ProcessWatcher` does not implement ICMP, meaning we're not tracking byte counts for any ICMP network traffic, just TCP and UDP
- Because this requires multiple parallel components to collect data at different times, there's always a possibility that short-lived processes don't get recorded. This is a problem with system/process in general, although it occurs here as well.


## Checklist



## How to test this PR locally

- pull down and build, set the `track_network_data` flag. Simply check incoming data for network traffic
- Alternatively, perform a large network operation, for example, download something with `wget`, and ensure the process is reported with the correct amount of traffic.

- Relates https://github.com/elastic/beats/issues/7461

## Performance 
I ran two tests with this metricset running under agent, then gathered a CPU profile via agent diagnostics. One test ran while downloading a 20GB file (the wikipedia archive at https://dumps.wikimedia.org/enwiki/20231120/), the other ran with just "ambient" network I/O.

In both cases, the `Tracker` thread is responsible for around 50% of agent's total CPU time. The majority of this time is spent in syscalls;

```
(pprof) tree Track
Active filters:
   focus=Track
Showing nodes accounting for 0.56s, 50.91% of 1.10s total
Showing top 80 nodes out of 114
----------------------------------------------------------+-------------
      flat  flat%   sum%        cum   cum%   calls calls% + context
----------------------------------------------------------+-------------
                                             0.16s 69.57% |   os.Readlink
                                             0.05s 21.74% |   internal/poll.ignoringEINTRIO
                                             0.01s  4.35% |   internal/poll.runtime_pollOpen
                                             0.01s  4.35% |   os.openFileNolog
     0.23s 20.91% 20.91%      0.23s 20.91%                | runtime/internal/syscall.Syscall6
```

This seems to be coming from packetbeat's `ProcessWatcher`:

<img width="380" alt="Screenshot 2023-11-30 at 1 31 29 PM" src="https://github.com/elastic/beats/assets/8418476/2d97c07d-adc5-4379-93df-030c59dd5264">


The processWatcher should cache results, so I'm a little baffled why this is using so much CPU time. Will continue investigating. 

EDIT: The issue has something to do with caching in the ProcessWatcher.  The `findProc` method will attempt to look up a address:port combination, and refresh its own mapping if it can't find it. That refresh is where we're getting hit. However, after some instrumenting, it looks like we're trying to look up the same address:port combination multiple times, which means that the internal map isn't getting updated when it should. Continuing to investigate.

EDIT2: issue here: https://github.com/elastic/beats/issues/37266
